### PR TITLE
Replace deprecated distutils

### LINKED ISSUE #2676: Replace distutils
We should replace our usage of `distutils` soonish:  `DeprecationWarning: The distutils package is deprecated and slated for removal in Python 3.12. Use setuptools or check PEP 632 for potential alternatives`

This is reported with Python 3.10, see https://github.com/cocotb/cocotb/pull/2675/checks?check_run_id=3364741779#step:25:199 for an example.

### PR DESCRIPTION
**Pull request by cmarqu**
distutils.version has been vendored in.

Closes #2676.

### CODE REVIEW

File: cocotb_build_libs.py [hunk 1]
```diff
@@ -6,17 +6,55 @@
...
 import textwrap
 
 from setuptools import Extension
-from distutils.spawn import find_executable
 from setuptools.command.build_ext import build_ext as _build_ext
-from distutils.file_util import copy_file
-from distutils.ccompiler import get_default_compiler
 from typing import List
 
+# vendored in from distutils/ccompiler.py:
```
Comment (ktbarrett) [inline cocotb_build_libs.py:18]: Are you sure this is necessary? I was under the impression that `setuptools` has a copy of `distutils` which is what gets imported when setup.py is run.

Comment (ktbarrett) [reply cocotb_build_libs.py:18]: See [this](https://github.com/pypa/setuptools/blob/0c302d232dd1a1f4ec531050532a84274edac2f4/_distutils_hack/__init__.py#L49) import hack in setuptools. I don't think we need to worry about vendoring in distutils code that is used in setup.py. We must only ensure that setuptools is imported first.

Comment (cmarqu) [reply cocotb_build_libs.py]: Ah yes, this works. Let's see if CI confirms...

**Comment by ktbarrett** [comment]
Surprise! Unexpected behavior! It doesn't seem to be omitting `cocotb/_vendor/distutil_version.py` from the coverage despite the `cocotb/_vendor/*` omit rule in `setup.cfg`.

**Comment by cmarqu** [comment]
> Surprise! Unexpected behavior! It doesn't seem to be omitting `cocotb/_vendor/distutil_version.py` from the coverage despite the `cocotb/_vendor/*` omit rule in `setup.cfg`.

Hard to believe that this should be a bug in Coverage.py...

**Comment by ktbarrett** [comment]
My experience tell me that there is probably a "very good reason" why it works this way.

**Comment by cmarqu** [comment]
Hmm, is it tox running this from a different path so that the relative location is different? Trying with a wildcard.

File: setup.cfg [hunk 2]
```diff
@@ -41,8 +41,8 @@ testpaths =
 
 [coverage:run]
```
Comment (ktbarrett) [inline setup.cfg:42]: I don't this this omit is in the right section.

**Comment by themperek** [comment]
We do not want a dependency on https://packaging.pypa.io/ ?

**Comment by ktbarrett** [comment]
@themperek We don't want a dependency and it won't work anyways since there is no `LooseVersion`. They only have classes for "clean" semantic versioning per PEP 440, which simulators tend to not follow.

File: setup.cfg (hunk 2 above)
Comment (cmarqu) [reply setup.cfg:42]: I'll try `coverage ... --debug=trace` later which should help in finding the problem, as per https://coverage.readthedocs.io/en/latest/cmd.html#diagnostics-coverage-debug

**Comment by cmarqu** [comment]
Huh, strange behavior. I added trace debugging, and locally, I'm getting output like
```
Source matching against trees <TreeMatcher ['/opt/tools/cocotb/cocotb']>
Omit matching: <FnmatchMatcher ['*/cocotb/config.py', '*/cocotb/_vendor/*']>
Not tracing '/tools/anaconda/2019.10-python3.7/Linux-x86_64/lib/python3.7/threading.py': falls outside the --source spec
[...]
Not tracing '/opt/tools/cocotb/tests/pytest/test_utils.py': falls outside the --source spec
Tracing '/opt/tools/cocotb/cocotb/_sim_versions.py'
Not tracing '/opt/tools/cocotb/cocotb/_vendor/__init__.py': is inside an --omit pattern
Not tracing '/opt/tools/cocotb/cocotb/_vendor/distutils_version.py': is inside an --omit pattern
```
but not in CI (also the first two lines are not there). This is with coverage 5.5 both times.

**Comment by ktbarrett** [comment]
@cmarqu Is there a reason to not bring this in for 1.6? We can figure out the coverage problem later if need be.

**Comment by cmarqu** [comment]
I'm fine with merging this as is.

**Comment by ktbarrett** [comment]
There is a failing test.

**Comment by cmarqu** [comment]
@themperek Do you see what's wrong here?: https://github.com/cocotb/cocotb/pull/2678/checks?check_run_id=3402399790#step:25:67
It's probably https://github.com/cocotb/cocotb/pull/2678/files#diff-1e9f10ea33562835a183dcbff5c9431b590b33053c7c59463cd83bfcba7dde2eR581-R584 that is behaving slightly different now.

**Comment by themperek** [comment]
> @themperek Do you see what's wrong here?: https://github.com/cocotb/cocotb/pull/2678/checks?check_run_id=3402399790#step:25:67
> It's probably https://github.com/cocotb/cocotb/pull/2678/files#diff-1e9f10ea33562835a183dcbff5c9431b590b33053c7c59463cd83bfcba7dde2eR581-R584 that is behaving slightly different now.

It looks like `config_vars` has changed. Maybe try `os.environ['LDSHARED']`? 
I will try to dig this.

**Comment by themperek** [comment]
Maybe also can try `ext.extra_link_args` here https://github.com/cocotb/cocotb/blob/master/cocotb_build_libs.py#L232

File: cocotb/_vendor/distutils_version.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
+#
+# distutils/version.py
+#
+# Implements multiple version numbering conventions for the
+# Python Module Distribution Utilities.
+#
+# $Id$
+#
```
Comment (ktbarrett) [inline cocotb/_vendor/distutils_version.py:8]: Does PSF require use to include the license with this file? That will also transitively reclassify the entire project as I think the PSF license is stricter than MIT. @imphil Are you more familiar with license issues?

Comment (imphil) [reply cocotb/_vendor/distutils_version.py:8]: No license needs to be included, but please prepend our standard (cocotb) SPDX header and add a copyright line following https://github.com/python/cpython/blob/main/LICENSE#L86-L88.

Comment (ktbarrett) [reply cocotb/_vendor/distutils_version.py:15]: I think I got that right.

**Comment by cmarqu** [comment]
Thanks for figuring all this out!

**Review by ktbarrett — APPROVED** [review]

**Comment by ktbarrett** [comment]
@cmarqu And thank you for pushing for Python 3.10 support in 1.6!

**Pull request merged**